{% macro isoform_result(isoform, indicate_preferred=False) %}
<div class="isoform">
  <h5 class="refseq">
    {{ isoform.refseq }}
    {% if indicate_preferred and isoform.is_preferred_isoform %}
      (preferred isoform)
    {% endif %}
  </h5>
  <div class="isoform-data">
      <div class="datum">Protein: {{ isoform.length }} residues</div>
      {% if isoform.cds_start and isoform.cds_end %}
      <div class="datum">CDS: {{ '{:,}'.format(isoform.cds_start) }}-{{ '{:,}'.format(isoform.cds_end) }}</div>
      {% endif %}
      {% if isoform.tx_start and isoform.tx_end %}
      <div class="datum">Transcription: {{ '{:,}'.format(isoform.tx_start) }}-{{ '{:,}'.format(isoform.tx_end) }}</div>
      {% endif %}
  </div>
  <div class="shortcuts">
    Views:
    <a href="{{ url_for('SequenceView:show', refseq=isoform.refseq) }}">
        Protein sequence
        ({{ isoform.sites | length }} PTM sites,
         {{ isoform.confirmed_mutations_count }} mutations)
    </a>
    {# Note: fetching interactors may be slow sometimes, although very useful #}
    {% set interactors = isoform.interactors_count %}
      <a href="{{ url_for('NetworkView:show', refseq=isoform.refseq) }}" class="{{ 'inactive' if not interactors }}">
      Site-specific PTM interaction network
      ({{ interactors or 0 }} PTM enzymes)
    </a>
  </div>
</div>
{% endmacro %}

{% macro gene_results(gene) %}
  <div class="result">
    <div class="result-top">
      <div class="matched-features">
        {% for feature, score in gene.matches.items() %}
          <span title="This gene was found based on {{ feature.pretty_name }} with {{ 1 / (1 + score) }} score">
        {{ feature.pretty_name }}
        </span>
        {% endfor %}
      </div>
      <h4 class="gene"><a href="{{ url_for('SequenceView:show', refseq=gene.preferred_isoform.refseq) }}">{{ gene.name }}</a></h4>
    </div>

    <div class="chrom">{{ gene.chrom }}</div>
    <div class="strand">{{ '+' if gene.strand else '-' }}</div>

    {% if gene.matched_isoforms %}
      <h6>Isoforms matching to query</h6>
      {% for isoform in gene.matched_isoforms %}
        {{ isoform_result(isoform, indicate_preferred=True) }}
      {% endfor %}
    {% endif %}

    {% if gene.preferred_isoform and gene.preferred_isoform not in gene.matched_isoforms %}
      <h6>Preferred isoform</h6>
      {{ isoform_result(gene.preferred_isoform) }}
    {% endif %}

    {% if gene.alternative_isoforms %}
      {% set alt_iso_cnt = gene.alternative_isoforms | length %}
      <input type="button" class="hidden btn btn-default show-alt js-shown"
       value="Show {{ alt_iso_cnt }} isoform{{ 's' if alt_iso_cnt > 1 }} more">
      <div class="alt-isoforms js-hidden">
        <h6>Alternative isoforms</h6>
        {% for isoform in gene.alternative_isoforms %}
          {{ isoform_result(isoform) }}
        {% endfor %}
      </div>
    {% endif %}
  </div>
{% endmacro %}

{% if gene %}
  {{ gene_results(gene) }}
{% endif %}
