{#
  Results should be in a form of a table where mappings are shown
  together with links to appropiate sites. Below the table information
  about positions that where not found in the database should be displayed
#}

<table class="table table-rowspan-hover">
  <thead>
  <tr>
    {% set columns_count = 9 %}
    <th>User input</th>
    <th>Gene name</th>
    <th>Refseq</th>
    <th>Ref aa</th>
    <th>Alt aa</th>
    <th>Position</th>
    <th>is PTM?</th>
    <th>Count</th>
    <th>Impact</th>
    <th>Link</th>
  </tr>
  </thead>
  {% for name, result_obj in results.items() %}
    <tbody>
    {% for result in result_obj.results %}
      {% set mutation = result.mutation %}
      <tr>
        {% if loop.first %}
          <td rowspan={{ loop.length }}>{{ result_obj.user_input }}</td>
          {% if result.type == 'proteomic' %}
          {# If the query was about one specific protein, it won't change #}
          <td rowspan={{ loop.length }}>
            <a href="{{ url_for('GeneView:show', gene_name=result.protein.gene.name)}}">
              {{ result.protein.gene.name }}
            </a>
          </td>
          {% endif %}
        {% endif %}
        {% if result.type != 'proteomic' %}
        <td>
          <a href="{{ url_for('GeneView:show', gene_name=result.protein.gene.name)}}">
            {{ result.protein.gene.name }}
          </a>
        </td>
        {% endif %}
        <td>
          <a href="{{ url_for('ProteinView:show', refseq=result.protein.refseq) }}">
            {{ result.protein.refseq }}
          </a>
        </td>
        <td>{{ result.ref }}</td>
        <td>{{ result.alt }}</td>
        <td>{{ result.pos }}</td>
        <td>{{ mutation.is_ptm() }}
        {#
          This might take advantage of precomputed PTMs for genomic mutations
          result.is_ptm if (result.is_ptm != None) else mutation.is_ptm
        #}</td>
        <td>{{ result_obj.count }}</td>
        {% set impact = mutation.impact_on_ptm() %}
        <td class="{{ impact or '' }}">
            {{ impact or '' }}
            {% if impact == 'network-rewiring' %}
                - motif {{ mutation.meta_MIMP.effect }}
            {% endif %}
        </td>
        <td>
            {% if mutation.is_confirmed %}
              {% for source in mutation.sources %}
                <a href="{{ url_for('ProteinView:show', refseq=result.protein.refseq, filters='Mutation.sources:in:' + source) }}#{{ result.pos }}{{ result.alt }}">
                mutation details ({{ source }}) â†’
              </a><br>
              {% endfor %}
            {% else %}
              no details
            {% endif %}
        </td>
      </tr>
    {% endfor %}
    </tbody>
  {% endfor %}
  <tfoot>
  {% if without_mutations %}
    <tr>
      <td colspan={{ columns_count }}>
        {{ without_mutations | length }} {{ 'queries' if without_mutations | length > 1 else 'query' }} without mutations found in the database.
        {% if without_mutations | length < 10 %}
          Those are:
          {{ without_mutations | map('trim') | join(', ') }}.
        {% endif %}
        <p>Some protein mutations are not possible to be caused by a Single Nucleotide Variation. Those are currently not shown in the table above.</p>

        <p>This might also indicate that given reference allele is not corresponding to given position or that there is a shift by one in your numeration.</p>
      </td>
    </tr>
  {% endif %}
  {% if badly_formatted %}
    <tr>
      <td colspan={{ columns_count }}>
        {{ badly_formatted | length }} {{ 'queries were' if badly_formatted | length > 1 else 'query was' }} not recognised by the system: {{ badly_formatted | map('trim') | join(', ') }}.
      </td>
    </tr>
    <tr>
      <td colspan={{ columns_count }}>
        "no details" means that given mutation is not present in any of datasets in this database. This might indicate somehow novel or rare mutation.
      </td>
    </tr>
  {% endif %}
  </tfoot>
</table>
